llvm/lib/Target/AMDGPU/AMDGPUCallingConv.td

```diff
@@ -185,9 +185,12 @@ def CSR_AMDGPU_NoRegs : CalleeSavedRegs<(add)>;
 // Calling convention for leaf functions
 def CC_AMDGPU_Func : CallingConv<[
   CCIfByVal<CCPassByVal<4, 4>>,
-  CCIfType<[i1], CCPromoteToType<i32>>,
   CCIfType<[i8, i16], CCIfExtend<CCPromoteToType<i32>>>,
 
+  CCIfType<[i1] , CCCustom<"CC_AMDGPU_Custom_I1">>,
+
+  CCIfType<[i1], CCPromoteToType<i32>>,
```

Comment:
Should only promote for inreg

Reply:
When CC_AMDGPU_Custom_I1 fails, i.e., an i1 arg is not allocated to an SGPR. If the i1 is not promoted to i32, we end up with a COPY from a 32b VGPR to a 64b VGPR, which is rejected by SIInstrInfo::copyPhysReg():

```
renamable $vgpr0_vgpr1 = COPY killed renamable $vgpr0, implicit $exec
```

Reply:
But I would still expect the first action to be covered by CCIfInReg

Reply:
The new CC function, CC_AMDGPU_Custom_I1, needs to have precedence over CCIfInReg. Otherwise, an "inreg i1" would be allocated to a 32b SGPR even if on a wavesize64 GPU.

Reply:
which you can do? Something like CCIfType<[i1], CCIfInReg<CCPromoteToType<i32>>

Reply:
I think the promotion needs to be done regardless of inreg. The problem is as follows:
```
foo(a bunch of args, i1 %argX, ...)
```
If the bunch of args (i1 or inreg) before argX takes all the SGPRs, then argX is going to get a VGPR regardless of inreg or not. If it's not promoted to i32, we end up with the COPY from 32b VGPR to 64b VGPR as shown above.

Reply:
No, inreg is changing the behavior of i1. The register decision is not a function of how many arguments are used.

What I want is "i1" -> SGPR pair.  "i1 inreg" -> promoted to 32-bit SGPR